ompi/mca/osc/ucx/osc_ucx_active_target.c

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2017. ALL RIGHTS RESERVED.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+
+#include "ompi/mca/osc/osc.h"
+#include "ompi/mca/osc/base/base.h"
+#include "ompi/mca/osc/base/osc_base_obj_convert.h"
+
+#include "osc_ucx.h"
+
+
+int ompi_osc_ucx_fence(int assert, struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*) win->w_osc_module;
+    int ret = OMPI_SUCCESS;
+    ucs_status_t status;
+
+    if (module->epoch_type.access != NONE_EPOCH &&
+        module->epoch_type.access != FENCE_EPOCH) {
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    if (assert & MPI_MODE_NOSUCCEED) {
+        module->epoch_type.access = NONE_EPOCH;
+    } else {
+        module->epoch_type.access = FENCE_EPOCH;
```

Comment:
IMO, the fence calls should be used in pairs. Standard says that:
"The call completes an RMA access epoch if it was preceded by another fence call and
the local process issued RMA communication calls on win between these two calls. The call
completes an RMA exposure epoch if it was preceded by another fence call and the local
window was the target of RMA accesses between these two calls. The call starts an RMA
access epoch if it is followed by another fence call and by RMA communication calls issued
between these two fence calls. The call starts an exposure epoch if it is followed by another
fence call and the local window is the target of RMA accesses between these two fence calls.
Thus, the fence call is equivalent to calls to a subset of post, start, complete, wait." (3.1; 441/8-16)

Does not it mean that you may use two fence calls without any assertions and the last one should complete access and exposure epochs?